packages/@aws-cdk/assertions/lib/match.ts

```diff
@@ -1,13 +1,242 @@
+
+  constructor(private readonly pattern: any[], options: ArrayMatchOptions = {}) {
+    super();
+    this.partial = options.partial ?? true;
+  }
+
+  public test(actual: any): MatchFailure[] {
+    if (!Array.isArray(actual)) {
+      return [{ path: [], message: `Expected type array but received ${getType(actual)}` }];
+    }
+    if (!this.partial && this.pattern.length !== actual.length) {
+      return [{ path: [], message: `Expected array of length ${this.pattern.length} but received ${actual.length}` }];
+    }
+
+    let patternIdx = 0;
+    let actualIdx = 0;
+
+    const failures: MatchFailure[] = [];
+    while (patternIdx < this.pattern.length && actualIdx < actual.length) {
+      let patternElement = this.pattern[patternIdx];
+      let matcher = Match.isMatcher(patternElement) ? patternElement : new LiteralMatch(patternElement);
+      const innerFailures = matcher.test(actual[actualIdx]);
+
+      if (!this.partial || innerFailures.length === 0) {
+        failures.push(...composeFailures(`[${actualIdx}]`, innerFailures));
+        patternIdx++;
+        actualIdx++;
+      } else {
+        actualIdx++;
+      }
+    }
+
+    for (; patternIdx < this.pattern.length; patternIdx++) {
+      const p = this.pattern[patternIdx];
+      const e = (Match.isMatcher(p) || typeof p === 'object') ? ' ' : ` [${p}] `;
+      failures.push({ path: [], message: `Missing element${e}at pattern index ${patternIdx}` });
+    }
+
+    return failures;
+  }
+}
+
+/**
+ * Options when initializing `ObjectMatch` class.
+ */
+export interface ObjectMatchOptions {
+  /**
+   * Whether the pattern should partially match with the target object.
+   * The target object can contain more keys than expected by the pattern.
+   * @default true
+   */
+  readonly partial?: boolean;
+}
+
+/**
+ * Match class that matches objects.
+ */
+export class ObjectMatch extends Match {
+  private readonly partial: boolean;
+  constructor(
+    private readonly pattern: {[key: string]: any},
+    options: ObjectMatchOptions = {}) {
+
+    super();
+    this.partial = options.partial ?? true;
+  }
+
+  public test(actual: any): MatchFailure[] {
+    if (typeof actual !== 'object' || Array.isArray(actual)) {
+      return [{ path: [], message: `Expected type object but received ${getType(actual)}` }];
+    }
+
+    const failures: MatchFailure[] = [];
+    if (!this.partial) {
+      for (const a of Object.keys(actual)) {
+        if (!(a in this.pattern)) {
+          failures.push({ path: [], message: `Unexpected key '${a}'` });
+        }
+      }
+    }
+
+    for (const [patternKey, patternVal] of Object.entries(this.pattern)) {
+      if (patternVal === ABSENT) {
+        if (patternKey in actual) {
+          failures.push({ path: [], message: `Expected key '${patternKey}' to be absent but present` });
+        }
+        continue;
+      }
+      if (!(patternKey in actual)) {
+        failures.push({ path: [], message: `Missing key '${patternKey}'` });
```

Comment:
I have a feeling it will help hierarchical printing and make more sense if we report this as:

```suggestion
        failures.push({ path: [`/${patternKey}`], message: `Missing key` });
```